test/integration/replicaset/replicaset_test.go

```diff
@@ -272,246 +279,650 @@ func TestAdoption(t *testing.T) {
 	defer closeFn()
-	podInformer := informers.Core().V1().Pods().Informer()
-	ns := framework.CreateTestingNamespace("rs-update-selector-to-remove-controllerref", s, t)
+	ns := framework.CreateTestingNamespace("test-replicaset-basics", s, t)
 	defer framework.DeleteTestingNamespace(ns, s, t)
+	labelMap := map[string]string{"foo": "bar"}
 	rs := newRS("rs", ns.Name, 2)
-	pod1 := newMatchingPod("pod1", ns.Name)
-	pod1.Labels["uniqueKey"] = "1"
-	pod2 := newMatchingPod("pod2", ns.Name)
-	pod2.Labels["uniqueKey"] = "2"
-	createRSsPods(t, clientSet, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{pod1, pod2}, ns.Name)
+	rs.Spec.Selector.MatchLabels = labelMap
+	rs.Spec.Template.Labels = labelMap
+	rss, _ := createRSsPods(t, c, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{})
+	rs = rss[0]
 
 	stopCh := make(chan struct{})
 	informers.Start(stopCh)
-	waitToObservePods(t, podInformer, 2)
 	go rm.Run(5, stopCh)
-	waitRSStable(t, clientSet, rs, ns.Name)
+	defer close(stopCh)
+	waitRSStable(t, c, rs)
+
+	scaleRS(t, c, rs, 3)
+	scaleRS(t, c, rs, 0)
+	scaleRS(t, c, rs, 2)
 
-	// change the rs's selector to match both pods
-	patch := `{"spec":{"selector":{"matchLabels": {"uniqueKey":"1"}},"template":{"metadata":{"labels":{"uniqueKey":"1"}}}}}`
-	rsClient := clientSet.Extensions().ReplicaSets(ns.Name)
-	rs, err := rsClient.Patch(rs.Name, types.StrategicMergePatchType, []byte(patch))
+	rsClient := c.Extensions().ReplicaSets(ns.Name)
+	rs, err := rsClient.Get(rs.Name, metav1.GetOptions{})
 	if err != nil {
-		t.Fatalf("Failed to patch replica set: %v", err)
+		t.Fatalf("failed to update rs %s: %v", rs.Name, err)
 	}
-	t.Logf("patched rs = %#v", rs)
-	// wait for the rs to create one more pod
-	if err := wait.Poll(10*time.Second, 60*time.Second, func() (bool, error) {
-		return verifyRemainingObjects(t, clientSet, ns.Name, 1, 3)
+
+	// Add an annotation to pod template to test rs's status does update without replicas change
+	rs.Annotations = map[string]string{"test": "annotation"}
+	newRS, err := rsClient.Update(rs)
+	if err != nil {
+		t.Fatalf("failed to update rs %s: %v", rs.Name, err)
+	}
+	if err := wait.PollImmediate(interval, timeout, func() (bool, error) {
+		curRS, err := rsClient.Get(rs.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		return curRS.Status.ObservedGeneration >= newRS.Generation, nil
 	}); err != nil {
-		t.Fatal(err)
+		t.Fatalf("Failed to verify .Status.ObservedGeneration has changed for rs %s: %v", rs.Name, err)
+	}
+
+	// Update 2 pods that RS created such that one is failed and one is being deleted
+	podClient := c.Core().Pods(ns.Name)
+	pods := getPods(t, podClient, labelMap)
+	if len(pods.Items) != 2 {
+		t.Fatalf("len(pods) = %d, want %d", len(pods.Items), 2)
+	}
+
+	failedPod := &pods.Items[0]
+	updatePod(t, podClient, failedPod, func(pod *v1.Pod) { pod.Status.Phase = v1.PodFailed })
+
+	deletingPod := &pods.Items[1]
+	deletePod(t, podClient, deletingPod, func(pod *v1.Pod) error {
```

Comment:
You don't need a deletePod wrapper, because there's no need to wrap in RetryOnConflict. As mentioned earlier, Delete does not check ResourceVersion so it would never return a conflict error.

You also don't need to GET the Pod just before deleting it. The only thing you pass to Delete is the Pod name, and that can't change, so you don't need to ask the server for the latest Pod value. You can just go straight to calling `Pods(pod.Namespace).Delete(pod.Name, ...)`.

Reply:
Oops! Yes, i remembered you mentioned that a deletion wrapper is unnecessary for `RetryOnConflict`. The wrapper was written when wrangling with `deletiontimestamp`, but was not removed when reverting back to `Delete`.